### DISCUSSION TITLE
Background refetch state machine

### DISCUSSION THREAD

**Discussion by benjaminpreiss** [statelyai/xstate#4934] [Q&A] [answered]
Dear all,

I am trying to build a txt record fetcher, that loads and analyzes a txt record of my choice.
Now I want this statemachine to be able to do a background check when I give a signal, but not leave the resulting state of analysis.

My sample implementaion looks like this and (ab)uses parallel states to implement the desired functionality.

<img width="1038" alt="Screenshot 2024-06-11 at 18 38 57" src="https://github.com/statelyai/xstate/assets/28349122/d094912f-1122-44eb-8f0b-3f5ce057694b">

Is this the correct way to implement such a system?

**Comment by davidkpiano** [comment, marked as answer]
Why do you not want to leave the resulting state of `analyze`?

In general, I would recommend _not_ transitioning between parallel states; try to keep them as orthogonal as possible. Remember that you can respond to the same event in each parallel state.

Without knowing the full details of your use-case, hopefully this serves as inspiration:
<img width="530" alt="CleanShot 2024-06-11 at 12 52 27@2x" src="https://github.com/statelyai/xstate/assets/1093738/65f75d18-0fa4-4f9c-b2c2-f041e87e3092">

**Reply by benjaminpreiss** [reply]
Wow, that was a quick answer... 🚀 Thanks for taking your time!

So, analyze invokes a function that does some logic on the loaded TXT records and depending on what that invoked function returns, the machine can transition to a variety of final states.

Generally speaking, I don't want the system to go into the "waiting" states on refetch. Only on the first fetch it should be in waiting state. That's because the UI shouldn't show the refetch that runs in the background, but only the final state until it is replaced by another final state...

Edit:

So, to summarize: I want the `waiting` state to only be active on the first load, and after that the system should alternate between `other` and `error`.

**Reply by benjaminpreiss** [reply]
I can also just create a state machine like this.

Then I would
- Store a machine snapshot upon each of the final states being reached in context on the key "past_state"
- Write some logic in the UI that creates a derived state from the stored past_state in the context and the actual state of the state machine

That would then be logic like this:
- machine: loading and past_state undefined -> loading
- machine: analyzing and past_state undefined -> analyzing
- machine: {some_final_state} and any past_state -> {some_final_state}

<img width="1038" alt="Screenshot 2024-06-11 at 19 10 12" src="https://github.com/statelyai/xstate/assets/28349122/d5a071b5-8075-43b3-9bd2-fb40bbe95705">
